### ISSUE TITLE
[QUESTION] Chapter 16: when to use padding

### ISSUE THREAD

**Issue by tinybucket** [ageron/handson-ml2#435]
Hi,

For Chapter 16 you introduced the idea of padding to ensure that the input sequences have the same length. In the answers in the appendix of the book you mentioned that keras will eventually support ragged tensors. Looking at the codes I suppose that keras already supports this feature

For Ch 16 Qn 8 you used ragged tensors and the train sizes are
X_train.shape=(10000, None)
y_train.shape=(10000, 1)

For Ch 16 Qn 9 **First version: a very basic seq2seq model you used padding so that the train sizes are**
X_train.shape=(10000, 18)
y_train.shape=(10000, 10)

My understanding is that RNN should be able to handle sequences of arbitary length. 
So why is it that we must use padding? and when should we use it?

**Comment by ageron** [comment]
Hi @tinybucket ,
Thanks for your interesting question!
Indeed, support for ragged tensors wasn't quite ready when I wrote the 2nd edition of the book, but now it works well (with still a few significant limitations, such as the fact that the [labels cannot be ragged tensors](https://colab.research.google.com/drive/1WZMVlwCL9rB_n32U-_HhN7SULZWh70pd?usp=sharing), and the fact that [TPUs are not supported yet](https://github.com/tensorflow/tensorflow/issues/27170), and [cuDNN cannot be used to accelerate `LSTM` or `GRU` when using ragged tensors](https://github.com/tensorflow/tensorflow/issues/48838)). I wrote the exercise solutions in chapter 16 after the 2nd edition was published, which is why I could use ragged tensors. 

Looking at [tf.keras's source code for the `LSTM` and `GRU` layers](https://github.com/tensorflow/tensorflow/blob/da96383680c0a320c9551e020c26132ae5ebb024/tensorflow/python/keras/layers/recurrent_v2.py#L1145), the `call()` method starts with this code:

```python
def call(self, inputs, mask=None, training=None, initial_state=None):
    inputs, row_lengths = backend.convert_inputs_if_ragged(inputs)
    is_ragged_input = (row_lengths is not None)
    [...]
```

Now let's look at [the `convert_inputs_if_ragged()` function's source code](https://github.com/tensorflow/tensorflow/blob/da96383680c0a320c9551e020c26132ae5ebb024/tensorflow/python/keras/backend.py#L6493):

```python
def convert_inputs_if_ragged(inputs):
  """Converts any ragged tensors to dense."""

  def _convert_ragged_input(inputs):
    if isinstance(inputs, ragged_tensor.RaggedTensor):
      return inputs.to_tensor()
    return inputs

  flat_inputs = nest.flatten(inputs)
  contains_ragged = py_any(
      isinstance(i, ragged_tensor.RaggedTensor) for i in flat_inputs)

  if not contains_ragged:
    return inputs, None

  inputs = nest.map_structure(_convert_ragged_input, inputs)
  # Multiple mask are not yet supported, so one mask is used on all inputs.
  # We approach this similarly when using row lengths to ignore steps.
  nested_row_lengths = math_ops.cast(flat_inputs[0].nested_row_lengths()[0],
                                     'int32')
  return inputs, nested_row_lengths
```

The most important lines are these two:

```python
if isinstance(inputs, ragged_tensor.RaggedTensor):
      return inputs.to_tensor()
```

Basically, any ragged tensor in an input batch is converted to a dense tensor automatically by calling `to_tensor()`, which automatically pads each sequence. So each sequence will end up being as long as the longest sequence in the batch. Notice that the function ends with `return inputs, nested_row_lengths`: it returns the dense tensor, and it also returns the length of each sequence in the batch.

In the [`LSTM` class's `call()` method](https://github.com/tensorflow/tensorflow/blob/da96383680c0a320c9551e020c26132ae5ebb024/tensorflow/python/keras/layers/recurrent_v2.py#L1171), the dense inputs and the row lengths are passed on to the `rnn()` function:

```python
last_output, outputs, states = backend.rnn(
          step,
          inputs, # <----- INPUTS ARE USED HERE
          initial_state,
          constants=None,
          go_backwards=self.go_backwards,
          mask=mask,
          unroll=self.unroll,
          input_length=row_lengths if row_lengths is not None else timesteps, # <----- ROW LENGTHS HERE
          time_major=self.time_major,
          zero_output_for_mask=self.zero_output_for_mask)
```

So, long story short, using ragged tensor is equivalent to padding each input batch to the size of its longest sequence (padding with zeros or empty strings, depending on the data type), and then masking the padding tokens during training based on each sequence's length.

In the solution to exercise 9, I called `to_tensor()` myself, to pad the sequences, but this would have been done automatically by Keras if I had not done it myself. I believe I had to do that because TF-Addons's seq2seq implementation did not support ragged tensors.

I hope this helps!

**Comment by tinybucket** [comment]
Hi @ageron,

Thanks for your answer, your explanation is very detailed and clear.
